server/src/main/java/org/elasticsearch/search/dfs/DfsPhase.java

```diff
@@ -39,71 +44,162 @@ public class DfsPhase {
+            return;
+        }
+
+        if (context.getProfilers() == null) {
+            executeKnnVectorQueryWithoutProfiling(context);
+        } else {
+            executeKnnVectorQueryWithProfiling(context);
+        }
+    }
+
+    private void executeKnnVectorQueryWithoutProfiling(SearchContext context) throws IOException {
+        KnnSearchBuilder knnSearch = context.request().source().knnSearch();
+        IndexSearcher searcher = new IndexSearcher(context.searcher().getIndexReader());
+        Query query = buildKnnVectorQuery(context);
+
+        TopDocs topDocs = searcher.search(query, knnSearch.k());
+
+        DfsKnnResults knnResults = new DfsKnnResults(topDocs.scoreDocs);
+        context.dfsResult().knnResults(knnResults);
+    }
+
+    private void executeKnnVectorQueryWithProfiling(SearchContext context) throws IOException {
+        long knnStartTime = System.nanoTime();
+
+        KnnSearchBuilder knnSearch = context.request().source().knnSearch();
+
+        Timer rewriteTimer = new Timer();
+        Timer createWeightTimer = new Timer();
+
+        IndexSearcher searcher = new IndexSearcher(context.searcher().getIndexReader()) {
+            @Override
+            public Query rewrite(Query original) throws IOException {
+                rewriteTimer.start();
+
+                try {
+                    return super.rewrite(original);
+                } finally {
+                    rewriteTimer.stop();
                 }
-                ParsedQuery query = searchExecutionContext.toQuery(knnVectorQueryBuilder);
+            }
 
-                TopDocs topDocs = searcher.search(query.query(), knnSearch.k());
-                DfsKnnResults knnResults = new DfsKnnResults(topDocs.scoreDocs);
-                context.dfsResult().knnResults(knnResults);
+            @Override
+            public Weight createWeight(Query query, ScoreMode scoreMode, float boost) throws IOException {
+                createWeightTimer.start();
+
+                try {
+                    return super.createWeight(query, scoreMode, boost);
+                } finally {
+                    createWeightTimer.stop();
+                }
             }
-        } catch (Exception e) {
-            throw new DfsPhaseExecutionException(context.shardTarget(), "Exception during dfs phase", e);
+        };
+
+        Query query = buildKnnVectorQuery(context);
+        TopScoreDocCollector tsdc = TopScoreDocCollector.create(knnSearch.k(), Integer.MAX_VALUE);
+        InternalProfileCollector ipc = new InternalProfileCollector(tsdc, "KnnVectorQuery", List.of());
+
+        try {
+            searcher.search(query, ipc);
+            DfsKnnResults knnResults = new DfsKnnResults(tsdc.topDocs().scoreDocs);
+            context.dfsResult().knnResults(knnResults);
+        } finally {
+            long knnTotalTime = System.nanoTime() - knnStartTime;
+            context.dfsResult()
+                .profileResult(
+                    new ProfileResult(
```

Comment:
Small comment, maybe it'd be nice to have a static method on `ProfileResult` like `createKnnResult`? It could hide some of these very profiler-specific details like creating a map the right keys, etc.

Reply:
@jtibshirani I agree that it should be moved, but I don't know if ProfileResult is the correct place to put it. If we're going to follow up this up with profiling for stats, would it make more sense to add this to part of the follow up and have a full custom `DfsProfileResult`? Or I could add one for this PR. What do you think?

Reply:
Sure, sounds good to do it in a follow-up.